Spigot-Server-Patches/0418-Ticket-level-per-player-view-distance-API.patch

```diff
@@ -0,0 +1,186 @@
+Subject: [PATCH] Ticket level per-player view distance API
+
+This PR implements the per-player view distance API at the ticket level.
+This somewhat clean solution comes from the result of many complicated ones (and research).
+The view-distance config value will determine the global distance (and set maximum) of normal view distance interactions with entities, mob-spawning and tracking.
+All chunks with player tickets within this range will also be served by the server, allowing shared view distance from other players without impacting performance.
+
+diff --git a/src/main/java/net/minecraft/server/ChunkMapDistance.java b/src/main/java/net/minecraft/server/ChunkMapDistance.java
+index 63a688725..879099b20 100644
+--- a/src/main/java/net/minecraft/server/ChunkMapDistance.java
++++ b/src/main/java/net/minecraft/server/ChunkMapDistance.java
+@@ -43,6 +43,7 @@ public abstract class ChunkMapDistance {
+     private final LongSet l = new LongOpenHashSet();
+     private final Executor m;
+     private long currentTick;
++    protected ObjectSet<EntityPlayer> getChunkPlayers(long i) {return (ObjectSet) ChunkMapDistance.this.c.get(i);} //Paper OBFHELPER
+
+     protected ChunkMapDistance(Executor executor, Executor executor1) {
+         executor1.getClass();
+@@ -352,7 +353,8 @@ public abstract class ChunkMapDistance {
+
+         private int e = 0;
+         private final Long2IntMap f = Long2IntMaps.synchronize(new Long2IntOpenHashMap());
+-        private final LongSet g = new LongOpenHashSet();
++        private final Long2ByteMap g = new Long2ByteOpenHashMap(); //Paper ticket level pp-view-dist
++        private Long2ByteMap getUpdateChunks() {return this.g;} //Paper OBFHELPER
+
+         protected c(int i) {
+             super(i);
+@@ -360,9 +362,9 @@ public abstract class ChunkMapDistance {
+         }
+
+         @Override
+-        protected void a(long i, int j, int k) {
+-            this.g.add(i);
+-        }
++        protected void a(long i, int j, int k, byte d) {
++            this.g.put(i, d);
++        } //Paper ticket level pp-view-dist
+
+         public void a(int i) {
+             ObjectIterator objectiterator = this.a.long2ByteEntrySet().iterator();
+@@ -372,20 +374,20 @@ public abstract class ChunkMapDistance {
+                 byte b0 = it_unimi_dsi_fastutil_longs_long2bytemap_entry.getByteValue();
+                 long j = it_unimi_dsi_fastutil_longs_long2bytemap_entry.getLongKey();
+
+-                this.a(j, b0, this.c(b0), b0 <= i - 2);
++                this.a(j, b0, this.c(b0), b0 <= i - 2, i); //Paper ticket level pp-view-dist
+             }
+
+             this.e = i;
+         }
+
+-        private void a(long i, int j, boolean flag, boolean flag1) {
+-            if (flag != flag1) {
++        private void a(long i, int j, boolean flag, boolean flag1, int vd) { //Paper ticket level pp-view-dist
++            if (flag != flag1 || (flag && flag1)) { //Paper ticket level pp-view-dist
```

Comment:
why did you add this condition to the branch

Reply:
consolidate ticket addition on view distance changes, especially increases

Reply:
what